File: project_euler/Problem083/Problem083.java
package project_euler;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

public class Problem083 {
    public void solution() throws IOException {
        String fileName = "matrix.txt";
        int border = 80;
        int[][] target = new int[border][border];
        try (BufferedReader handle = new BufferedReader(new InputStreamReader(new FileInputStream(fileName)))) {
            String line;
            int index = 0;
            while (null != (line = handle.readLine())) {
                String[] elements = line.split(" ");
                for (int i = elements.length - 1; i >= 0; --i) {
                    target[index][i] = Integer.parseInt(elements[i]);
                }
                ++index;
            }
        }

        int length = border * border;
        int[][] edges = new int[length][length];
        for (int i = 0; i < length; ++i) {
            Arrays.fill(edges[i], 0);
        }
        for (int i = 0; i < border; ++i) {
            for (int j = 0; j < border; ++j) {
                if (i > 0) {
                    edges[i * border + j][(i - 1) * border + j] = target[i - 1][j];
                }
                if (j > 0) {
                    edges[i * border + j][i * border + j - 1] = target[i][j - 1];
                }
                if (i < border - 1) {
                    edges[i * border + j][(i + 1) * border + j] = target[i + 1][j];
                }
                if (j < border - 1) {
                    edges[i * border + j][i * border + j + 1] = target[i][j + 1];
                }
            }
        }

        int[] ways = new int[length];
        search(edges, length, ways);
        System.out.println(ways[length - 1] + target[0][0]);
    }

    private void search(int[][] edges, int length, int[] ways) {
        boolean[] visited = new boolean[length];
        Arrays.fill(visited, false);
        Queue<Heap> queue = new PriorityQueue<>(Comparator.comparingInt(p -> p.weight));
        queue.offer(new Heap(0, 0));
        while (!queue.isEmpty()) {
            Heap node = queue.poll();
            int index = node.index;
            int weight = node.weight;
            if (visited[index]) {
                continue;
            }

            ways[index] = weight;
            visited[index] = true;
            for (int i = 0; i < length; ++i) {
                if (0 != edges[index][i] && !visited[i]) {
                    queue.offer(new Heap(i, weight + edges[index][i]));
                }
            }
        }
    }

    private static class Heap {
        public int index;
        public int weight;

        public Heap(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }
}

File: project_euler/Problem083/Problem083.js
function Problem083() {
    let fileName = 'matrix.txt'
    let border = 80
    let target = Array(border)
    for (let i = 0; i < border; ++i) {
        target[i] = Array(border)
    }

    const file = require('n-readlines')
    let handle = new file(fileName)
    let line
    let index = 0
    while (line = handle.next()) {
        let elements = line.toString().split(' ')
        for (let [i, element] of elements.entries()) {
            target[index][i] = Number(element)
        }
        ++index
    }

    let length = border * border
    let edges = Array(length)
    for (let i = 0; i < length; ++i) {
        edges[i] = Array(length).fill(0)
    }
    for (let i = 0; i < border; ++i) {
        for (let j = 0; j < border; ++j) {
            if (i > 0) {
                edges[i * border + j][(i - 1) * border + j] = target[i - 1][j]
            }
            if (j > 0) {
                edges[i * border + j][i * border + j - 1] = target[i][j - 1]
            }
            if (i < border - 1) {
                edges[i * border + j][(i + 1) * border + j] = target[i + 1][j]
            }
            if (j < border - 1) {
                edges[i * border + j][i * border + j + 1] = target[i][j + 1]
            }
        }
    }

    let ways = Array(length)
    search(edges, length, ways)
    console.log(ways[length - 1] + target[0][0])
}

function search(edges, length, ways) {
    let visited = Array(length).fill(false)
    let queue = [new Heap(0, 0)]
    while (queue.length > 0) {
        let node = queue[0]
        let index = node.index
        let weight = node.weight
        dequeue(queue)
        if (visited[index]) {
            continue
        }

        ways[index] = weight
        visited[index] = true
        for (let i = 0; i < length; ++i) {
            if (edges[index][i] && !visited[i]) {
                enqueue(i, weight + edges[index][i], queue)
            }
        }
    }
}

function enqueue(index, weight, queue) {
    let child = queue.length
    queue.push(new Heap(0, 0))
    while (child > 0) {
        let parent = (child - 1) >>> 1
        if (queue[parent].weight <= weight) {
            break
        }
        queue[child].index = queue[parent].index
        queue[child].weight = queue[parent].weight
        child = parent
    }
    queue[child].index = index
    queue[child].weight = weight
}

function dequeue(queue) {
    let pivot = queue.pop()
    let parent = 0
    while (true) {
        let child = (parent << 1) + 1
        if (child >= queue.length) {
            break
        }

        if (child < queue.length - 1 && queue[child].weight > queue[child + 1].weight) {
            ++child
        }
        if (queue[child].weight >= pivot.weight) {
            break
        }
        queue[parent].index = queue[child].index
        queue[parent].weight = queue[child].weight
        parent = child
    }
    if (queue.length) {
        queue[parent].index = pivot.index
        queue[parent].weight = pivot.weight
    }
}

function Heap(index, weight) {
    this.index = index
    this.weight = weight
}

module.exports = {
    Problem083 : Problem083
};
